orchagent/copporch.cpp

```diff
@@ -122,27 +185,118 @@ const string default_trap_group = "default";
     m_vidToRidTable(std::unique_ptr<Table>(new Table(m_asic_db.get(), "VIDTORID"))),
-    m_trap_counter_manager(HOSTIF_TRAP_COUNTER_FLEX_COUNTER_GROUP, StatsMode::READ, HOSTIF_TRAP_COUNTER_POLLING_INTERVAL_MS, false)
+    m_trap_counter_manager(HOSTIF_TRAP_COUNTER_FLEX_COUNTER_GROUP, StatsMode::READ, HOSTIF_TRAP_COUNTER_POLLING_INTERVAL_MS, false),
+    m_trapCapabilityTable(std::unique_ptr<Table>(new Table(m_state_db.get(), STATE_COPP_TRAP_CAPABILITY_TABLE_NAME))),
+    m_trapTable(std::unique_ptr<Table>(new Table(m_state_db.get(), STATE_COPP_TRAP_TABLE_NAME)))
 {
     SWSS_LOG_ENTER();
     auto intervT = timespec { .tv_sec = FLEX_COUNTER_UPD_INTERVAL , .tv_nsec = 0 };
     m_FlexCounterUpdTimer = new SelectableTimer(intervT);
     auto executorT = new ExecutableTimer(m_FlexCounterUpdTimer, this, "FLEX_COUNTER_UPD_TIMER");
     Orch::addExecutor(executorT);
 
+    /* Query SAI for supported trap IDs and publish to STATE_DB */
+    publishTrapIdsCapability();
+
     initDefaultHostIntfTable();
     initDefaultTrapGroup();
     initDefaultTrapIds();
+
 };
 
+bool CoppOrch::isTrapIdSupported(sai_hostif_trap_type_t trap_id) const
+{
+    return supported_trap_ids.find(trap_id) != supported_trap_ids.end();
+}
+
+void CoppOrch::updateTrapOperStatus(sai_hostif_trap_type_t trap_type, const string& hw_status)
+{
+    SWSS_LOG_ENTER();
+
+    string trap_name = get_trap_name_by_type(trap_type);
+    if (trap_name.empty())
+    {
+        SWSS_LOG_ERROR("Failed to get trap name for type %d", trap_type);
+        return;
+    }
+
+    // Update or add the hw_status field in the table
+    vector<FieldValueTuple> hwStatusFvs;
+    hwStatusFvs.emplace_back("hw_status", hw_status);
+    m_trapTable->set(trap_name, hwStatusFvs);
+}
+
+// Query SAI for trap IDs capability and publish to the COPP_TRAP_CAPABILITY_TABLE
+void CoppOrch::publishTrapIdsCapability()
+{
+    SWSS_LOG_ENTER();
+
+    sai_s32_list_t enum_values_capability;
+
+    supported_trap_ids.clear();
+
+    const auto* meta = sai_metadata_get_attr_metadata(SAI_OBJECT_TYPE_HOSTIF_TRAP, SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE);
+    if (!meta || !meta->isenum) 
+    {
+        SWSS_LOG_THROW("sai_metadata_get_attr_metadata for SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE Failed");
+        return;    
+    }
+
+    vector<int32_t> values_list(meta->enummetadata->valuescount);    
+    enum_values_capability.count = static_cast<uint32_t>(values_list.size());
+    enum_values_capability.list = values_list.data();
+
+    sai_status_t status = sai_query_attribute_enum_values_capability(gSwitchId,
+                                                                     SAI_OBJECT_TYPE_HOSTIF_TRAP,
+                                                                     SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE,
+                                                                     &enum_values_capability);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_WARN("SAI capability query for SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE is failed,"
```

Comment:
This can be changed to NOTICE as there could be many cases (like DPU) where this can return NOT_SUPPORTED